src/sentry/snuba/metrics.py

```diff
@@ -1112,16 +1077,167 @@ def get_series(self, projects: Sequence[Project], query: QueryDefinition) -> dic
+                groupby_tags = [
+                    resolve_tag_key(field) if field not in _ALLOWED_GROUPBY_COLUMNS else field
+                    for field in query.groupby
+                ]
+
+                # Dictionary that contains the conditions that are required to be added to the where
+                # clause of the second query. In addition to filtering down on the tuple combination
+                # of the fields in the group by columns, we need a separate condition for each of
+                # the columns in the group by with their respective values so Clickhouse can
+                # filter the results down before checking for the group by column combinations.
+                ordered_tag_conditions = {
+                    col: list({data_elem[col] for data_elem in initial_query_results["data"]})
+                    for col in groupby_tags
+                }
+                ordered_tag_conditions[tuple(col for col in groupby_tags)] = [
+                    tuple(data_elem[col] for col in groupby_tags)
+                    for data_elem in initial_query_results["data"]
+                ]
+
+                for entity, queries in snuba_queries.items():
+                    # This loop has constant time complexity as it will always have a maximum of
+                    # three queries corresponding to the three available entities
+                    # ["metrics_sets", "metrics_distributions", "metrics_counters"]
+                    snuba_query = queries["totals"]
+
+                    # If query is grouped by project_id, then we should remove the original
+                    # condition project_id cause it might be more relaxed than the project_id
+                    # condition in the second query
+                    where = []
+                    if "project_id" in groupby_tags:
+                        for condition in snuba_query.where:
+                            if not (
+                                isinstance(condition.lhs, Column)
+                                and condition.lhs.name == "project_id"
+                            ):
+                                where += [condition]
+
+                    # Adds the conditions obtained from the previous query
+                    for condition_key, condition_value in ordered_tag_conditions.items():
+                        lhs_condition = (
+                            Function("tuple", [Column(col) for col in condition_key])
+                            if isinstance(condition_key, tuple)
+                            else Column(condition_key)
+                        )
+                        where += [
+                            Condition(lhs_condition, Op.IN, Function("tuple", condition_value))
+                        ]
+                    snuba_query = snuba_query.set_where(where)
+
+                    snuba_query_res = raw_snql_query(
+                        snuba_query, use_cache=False, referrer="api.metrics.totals"
```

Comment:
Should we use a different referrer here than for the initial query?

Reply:
Yeah that makes sense!